### ISSUE TITLE
Cannot drop database due to a running 'session'

### ISSUE THREAD

**Issue by nriser** [ga-wdi-boston/full-stack-project#822]
I'm trying to drop my database (that I made for practice) and start a new one. 

I've tried the ```bin/rails db:drop``` command but I keep getting the error below that says that there is a session that's using the database: 

```
$ bin/rails db:drop
PG::ObjectInUse: ERROR:  database "flashcard-api_development" is being accessed by other users
DETAIL:  There is 1 other session using the database.
: DROP DATABASE IF EXISTS "flashcard-api_development"
Couldn't drop database 'flashcard-api_development'
rails aborted!
ActiveRecord::StatementInvalid: PG::ObjectInUse: ERROR:  database "flashcard-api_development" is being accessed by other users
DETAIL:  There is 1 other session using the database.
: DROP DATABASE IF EXISTS "flashcard-api_development"
/Users/nriser/wdi/projects/project-flashcard/flashcard-api/bin/rails:9:in `require'
/Users/nriser/wdi/projects/project-flashcard/flashcard-api/bin/rails:9:in `<top (required)>'
/Users/nriser/wdi/projects/project-flashcard/flashcard-api/bin/spring:15:in `<top (required)>'
bin/rails:3:in `load'
bin/rails:3:in `<main>'
PG::ObjectInUse: ERROR:  database "flashcard-api_development" is being accessed by other users
DETAIL:  There is 1 other session using the database.
/Users/nriser/wdi/projects/project-flashcard/flashcard-api/bin/rails:9:in `require'
/Users/nriser/wdi/projects/project-flashcard/flashcard-api/bin/rails:9:in `<top (required)>'
/Users/nriser/wdi/projects/project-flashcard/flashcard-api/bin/spring:15:in `<top (required)>'
bin/rails:3:in `load'
bin/rails:3:in `<main>'
Tasks: TOP => db:drop:_unsafe
(See full trace by running task with --trace)
```
 I rebooted rails server multiple times and kept getting the same thing above.

I became desperate and ran ```bin/rake db:migrate VERSION=0``` (as you can see below) but my schema still has the two tables that I made. I expected the schema to revert back to the original template before I made the two tables. 

```
~/wdi/projects/project-flashcard/flashcard-api (master)
$ bin/rake db:migrate VERSION=0
Running via Spring preloader in process 73626
== 2 CreateExamples: reverting ================================================
-- drop_table(:examples, {})
   -> 0.0843s
== 2 CreateExamples: reverted (0.0904s) =======================================

== 1 CreateUsers: reverting ===================================================
-- drop_table(:users, {})
   -> 0.0015s
== 1 CreateUsers: reverted (0.0018s) ==========================================

```
I also checked in psql and the database is still there (although the users table is gone):
```
flashcard-api_development=# \d
                 List of relations
 Schema |         Name         |   Type   | Owner  
--------+----------------------+----------+--------
 public | ar_internal_metadata | table    | nriser
 public | authors              | table    | nriser
 public | authors_id_seq       | sequence | nriser
 public | schema_migrations    | table    | nriser
 public | words                | table    | nriser
 public | words_id_seq         | sequence | nriser
(6 rows)
```

my schema:

```ruby
ActiveRecord::Schema.define(version: 20170503043529) do

  # These are extensions that must be enabled in order to support this database
  enable_extension "plpgsql"

  create_table "authors", force: :cascade do |t|
    t.string   "given_name",  null: false
    t.string   "family_name", null: false
    t.datetime "created_at",  null: false
    t.datetime "updated_at",  null: false
  end

  create_table "words", force: :cascade do |t|
    t.string   "word",       null: false
    t.string   "definition", null: false
    t.datetime "created_at", null: false
    t.datetime "updated_at", null: false
    t.integer  "author_id"
    t.index ["author_id"], name: "index_words_on_author_id", using: :btree
  end

  add_foreign_key "words", "authors"
end
```

I've checked online to find solutions to successfully drop a database, but I want to check with a consultant first before I break something. Although, it seemed to have deleted the user table that came with the template. 

Should I just re-download the full-stack api tempalate? Would there be issues with connecting with heroku?

**Comment by MicFin** [comment]
Is your server running?  You should shut it down before you try to drop the database.

**Comment by nriser** [comment]
@MicFin Thank you for your reply. Yes it's running and I did shut it down a couple of times but kept getting the same error. Seems too late now because I seemed to have messed up the template.

**Comment by MicFin** [comment]
Nothing can break the template! ;) 

In the error message it says `There is 1 other session using the database.` . If the server is running then that is the problem.  You should shut it down before you try to drop the database with [these commands](https://github.com/ga-wdi-boston/rails-api-template#reset-database-without-dropping)

**Comment by nriser** [comment]
I ran ```bin/rake db:migrate VERSION=0``` but my schema still has the two tables that I made (authors and words). I expected the schema to revert back to the original template before I made the two tables. 

I suppose I could run ```bin/rake db:migrate db:seed db:examples ``` to re-migrate to get the users and the examples tables but it worries me that my two tables are still there even after I ran ```bin/rake db:migrate VERSION=0``` :(

**Comment by MicFin** [comment]
Sounds like something to look at with fresh eyes tomorrow 💯

**Comment by nriser** [comment]
May I request a 1 on 1 with you sometime tomorrow? @MicFin

**Comment by MicFin** [comment]
Let's see if you can solve this remotely.  If your server is turned off and you run `bin/rake db:migrate VERSION=0`...
What does the ID of your schema say? 
What is the list of IDs for your migration files (the file names)?

**Comment by nriser** [comment]
@MicFin I ran ```bin/rake db:migrate VERSION=0``` last night and the version changed to 20170503043529

**Comment by MicFin** [comment]
Try `bin/rake db:rollback`.  That will rollback 1 migration at a time.   You should be able to do it a few times and then be back to 0.

**Comment by nriser** [comment]
@MicFin I ran that command 5 times just now but the version stays the same at 20170503043529. no change. My words and authors tables are still in the schema.

**Comment by nriser** [comment]
Thank you @MicFin for your help. 

The issue was that I built the two tables on a 'testing' branch, but ran db:drop/db:rollback/db:migrate VERSION=0 on my master branch. When I ran those commands in master, it couldn't find the files/changes that were actually done on the 'testing' branch, so rails kept giving me errors. To solve this problem, I went back to master and ran ```git checkout -- filename```, which reverted the schema back to version 2 on master.
